File: sdk/android/frak-sdk-ui/src/main/kotlin/id/frak/sdk/ui/SharingResult.kt
package id.frak.sdk.ui

import id.frak.sdk.core.FrakError

/**
 * How a sharing session ended. A session can produce several; the callback
 * reports only the most significant: install > shared/copied > dismissed.
 */
public sealed interface SharingResult {
    /**
     * Stable discriminator, one per arm. A `when` over [Kind] with an `else` survives a new arm;
     * a `when` over the hierarchy does not. [Kind.wireValue] is spelled identically on iOS.
     */
    public val kind: Kind

    public enum class Kind(
        public val wireValue: String,
    ) {
        SHARED("shared"),
        COPIED("copied"),
        INSTALL_STARTED("installStarted"),
        DISMISSED("dismissed"),
        FAILED("failed"),
    }

    public class Shared(
        public val link: String,
    ) : SharingResult {
        override val kind: Kind get() = Kind.SHARED
    }

    public class Copied(
        public val link: String,
    ) : SharingResult {
        override val kind: Kind get() = Kind.COPIED
    }

    /**
     * The user asked to install; the sheet took them to the wallet's install page, or to the
     * store with no identity to hand it. Informational only — do not call
     * [id.frak.sdk.AppLinkApi.openFrakApp] again in response, and it doesn't mean anything was
     * installed.
     */
    public object InstallStarted : SharingResult {
        override val kind: Kind get() = Kind.INSTALL_STARTED
    }

    public object Dismissed : SharingResult {
        override val kind: Kind get() = Kind.DISMISSED
    }

    public class Failed(
        public val error: FrakError,
    ) : SharingResult {
        override val kind: Kind get() = Kind.FAILED
    }
}

/** Higher wins. */
internal val SharingResult.significance: Int
    get() =
        when (this) {
            is SharingResult.Failed -> 0
            is SharingResult.Dismissed -> 1
            is SharingResult.Shared, is SharingResult.Copied -> 2
            is SharingResult.InstallStarted -> 3
        }

File: sdk/ios/Sources/FrakSDKUI/SharingResult.swift
import FrakSDK

/// How a sharing sheet ended.
public enum SharingResult: Sendable {
    case shared(link: String)
    case copied(link: String)
    /// The user asked to install and the sheet took them to the wallet's install page (or, with
    /// no identity to hand it, to the store). Informational only — do not call `openFrakApp()`
    /// in response; the sheet owns the step from here. Does not mean anything installed.
    case installStarted
    /// `openFrakApp()` answered `.openedApp` — an install this SDK observed and handed off, not a
    /// claim the backend has linked it. A merchant who needs "linked" reads that from the backend.
    case walletOpened
    case dismissed
    case failed(FrakError)

    /// Stable discriminator, one per case. A `switch` over ``Kind`` with a `default` survives a
    /// new case; an exhaustive `switch` over the result does not. Spelled identically on Android.
    public enum Kind: String, Sendable, Hashable, CaseIterable {
        case shared
        case copied
        case installStarted
        case walletOpened
        case dismissed
        case failed
    }

    public var kind: Kind {
        switch self {
        case .shared: return .shared
        case .copied: return .copied
        case .installStarted: return .installStarted
        case .walletOpened: return .walletOpened
        case .dismissed: return .dismissed
        case .failed: return .failed
        }
    }
}

extension SharingResult {
    /// One session can produce several outcomes; the caller is told the most significant one.
    var significance: Int {
        switch self {
        case .failed: 0
        case .dismissed: 1
        case .shared, .copied: 2
        case .installStarted: 3
        case .walletOpened: 4
        }
    }
}
